webapp/src/js/services/enketo-translation.js

```diff
@@ -1,294 +1,53 @@
-              if (attrs.appearance) {
-                attrs.appearance += ' allow-new';
-              } else {
-                attrs.appearance = 'allow-new';
-              }
-            }
-            var input = new N('input', attrs);
-            input.append(new N('label', translationFor(schema.type, 'field', f)));
-            fields.push(input);
-          });
-          return fields;
-        };
-
-        if(extras) {
-          body.append(groupFor(principle, principle.type));
-        } else {
-          body.append(fieldsFor(principle, principle.type));
-        }
-
-        _.each(extras, function(extra, mapping) {
-          var group = groupFor(extra, mapping);
-          group.prepend(new N('label', translationFor('contact.type', extra.type, 'new')));
-          body.append(group);
-        });
-
-        return body;
-      }());
-      root.append(body);
-
-      return root.xml();
+    const withElements = nodes => {
+      return Array.from(nodes)
+        .filter(node => node.nodeType === Node.ELEMENT_NODE);
     };
 
-    self.getHiddenFieldList = function(model) {
-      model = $.parseXML(model).firstChild;
-      return model && withElements(model.childNodes)
-        .filter(function(n) {
-          var attr = n.attributes.getNamedItem('tag');
-          return attr && attr.value === 'hidden';
-        })
-        .map(function(n) {
-          return n.nodeName;
-        })
-        .value();
+    const findChildNode = function(root, childNodeName) {
+      return withElements(root.childNodes)
+          .find(node => node.nodeName === childNodeName);
     };
 
     var nodesToJs = function(data, repeatPaths, path) {
       repeatPaths = repeatPaths || [];
       path = path || '';
       var result = {};
-      withElements(data)
-        .each(function(n) {
-          var dbDocAttribute = n.attributes.getNamedItem('db-doc');
-          if (dbDocAttribute && dbDocAttribute.value === 'true') {
-            return;
-          }
+      withElements(data).forEach(function(n) {
+        var dbDocAttribute = n.attributes.getNamedItem('db-doc');
+        if (dbDocAttribute && dbDocAttribute.value === 'true') {
+          return;
+        }
 
-          var typeAttribute = n.attributes.getNamedItem('type');
-          var updatedPath = path + '/' + n.nodeName;
-          var value;
+        var typeAttribute = n.attributes.getNamedItem('type');
+        var updatedPath = path + '/' + n.nodeName;
+        var value;
 
-          var hasChildren = withElements(n.childNodes).size().value();
-          if(hasChildren) {
-            value = nodesToJs(n.childNodes, repeatPaths, updatedPath);
-          } else if (typeAttribute && typeAttribute.value === 'binary') {
-            // this is attached to the doc instead of inlined
-            value = '';
-          } else {
-            value = n.textContent;
-          }
+        var hasChildren = withElements(n.childNodes).length > 0;
```

Comment:
Whoa, this file should have been a PR of it's own :)
Can `withElements(n.childNodes).length` be negative? Joking, no need to update.